From 73b12b82d74220b0152c98ce6a5305a48c23b57f Mon Sep 17 00:00:00 2001
From: Serguei Katkov <serguei.i.katkov@intel.com>
Date: Fri, 24 May 2013 18:32:18 +0700
Subject: Dalvik: Invalid assert statement in LoopRegisterUsage

BZ: 110740

Invalid assert statement suggests that defWhere field always allocated.
However it is not true. It is allocated only if this MIR has uses.
The patch correct assert to take into account the case above.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I829762c2a279e881612a953ea22164afb151da75
Orig-MCG-Change-Id: I55bc61c77af9e4cb6042101a2b4a421b1e69131d
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/LoopRegisterUsage.cpp |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/vm/compiler/LoopRegisterUsage.cpp b/vm/compiler/LoopRegisterUsage.cpp
index bf7eabe..9a8700f 100644
--- a/vm/compiler/LoopRegisterUsage.cpp
+++ b/vm/compiler/LoopRegisterUsage.cpp
@@ -448,7 +448,8 @@ static bool similarInstructions (MIR *first, MIR *second, const std::map<MIR *,
     MIR **defWhereSecond = ssaRepSecond->defWhere;
 
     //Paranoid
-    assert (defWhereFirst != 0 && defWhereSecond != 0);
+    assert (numUsesFirst <= 0 || defWhereFirst != 0);
+    assert (numUsesSecond <= 0 || defWhereSecond != 0);
 
     //Let's iterate through it
     for (int i = 0; i + startUseIndexFirst < numUsesFirst; i++)
-- 
1.7.4.1

